### ISSUE TITLE
When is the tentative release date of Alpha 3.0?

### ISSUE THREAD

**Issue by vaibhav-kaushal** [qcubed/framework#601]
@spekary You have been adding more and more **features** into alpha 3.0. But when is the plan to release it? Since a lot of things are moving in and out, I am not willing to even go ahead with the documentation bit. 

Also, my current project needs some of the overhauls that alpha branch provides (which includes Bootstrap). Also, it is in alpha. Unless it gets to beta stage where the feature set is at least frozen, I cannot make a move to v3.0. So I am waiting and am in a fix. A release date would be really nice to know.

**Comment by olegabr** [comment]
+1

**Comment by olegabr** [comment]
@spekary  did a lot of excellent improvements. the 3.0 release should be codenamed "qcubed 3.0 SPEKARY" :-)
But @vaibhav-kaushal is right. we need to release at some point. @spekary, please, provide some must-have features list that you are still planning for the 3.0

**Comment by vaibhav-kaushal** [comment]
If there were a codename that had to be chose, I would settle for ‘Shannon’ (his real name). It has been made possible by him by a very large extent. 

However, I think the time has come that we stop putting in new features and freeze and test the existing one.

**Comment by spekary** [comment]
Great questions!

I appreciate the recognition. No codename please though. Most of this work I have been doing as I push to creating some consumer level products and bringing on other developers, as well as add Bootstrap support.

And all, I realize I have been making a lot of changes that are potentially destabilizing. But, the code base is so old that I think it needed a core update in many areas, and so I am looking to doing this now so we can test and get through beta and have a solid platform for the future. 

In particular, as I added Bootstrap support, some big problems with the underlying code generation, least of which was the ability to understand and support it. A good chunk of what I have been doing lately has been more refactoring than new features. 

The initial bootstrap plugin is up in the repositories. Its a start, but can already do a navbar, a carousel and really nice looking forms. I am hoping others will add to it ;-) You can see the code working at www.ryaa.org

Here is what I would like to do before getting to beta and calling it code complete:
1. There is an 'eval' in qcubed.js with a todo note to try to remove it. I needed to do some changes to how javascript code was executed, and it kind of ballooned from there. I am now working on infrastructure to be able to execute most of the javascript in qcubed without going through the eval. I don't intend to take out the eval yet, but just have the architecture there so that developers would only need it rarely. As part of this, I need to refactor the spaghetti in RenderEnd and RenderAjax to make it easier to understand and build on in the future.
2. I mentioned a change in how Persistence works. I need #1 to be able to do this step. This change will mean that you can do this:
```
$objControl->SaveState = true
```
and the object will have the ability to restore its *visible* state. I don't know if anyone uses the current persistence mechanism, but its a memory pig, and its design goals were to be a kind of caching of controls with complex data. Its not really the right way to cache that data, and the implementation is cumbersome. The new version will only save the minimum amount of info to restore the visual state, and not the underlying data, which I think is what most people need. It also makes it easy to save the state of all the controls in the whole application for a particular user, and then restore that state when they log back in, so they feel like nothing changed between logins.
3. I need to do a few tweaks to the new formObjChanged implementation just to simplify it a bit.
4. The final effort will be to resolve the name of MetaControls. I would like to change it to ModelController. The nice thing about this is that the 'mct' prefix can stay the same, so most user code will not need to change much. The name is reflective of what it does, using the language of MVC. This work will need a change of name to the new MetaParams and MetaControl designer dialog. But again, this is all name changing an not really different code.

QCubed is a really nice architecture, in that it breaks out logical functions well. Our "Model" actually incorporates the idea of a ViewModel, in that the generated base class is the Model and the subclass is the ViewModel. Our Controller is split between a customizable ModelController(the current MetaControl), and a ViewController (the current FormObj). The Views are the templates.

There are then some things to do during beta:
1. Write documentation
2. Fix bugs in the examples code
3. Update the website
4. Possibly write visual unit tests using Codeception? The tests could run through the examples code and make sure they are all working as expected. The problem with this idea is that to really get a good test, it needs to be run on a Firefox browser because of the javascript, and so making it part of the travis build would not be an option. 
5. Screencast tutorials? Introduction, Architectural Overview, Installation, Codegeneration, Example application build from start to finish would all be great.
6. Someone mentioned a deployment strategy. That might be just a documentation piece, or an actual tool. I think it would be an excellent addition.

When we go to RC, we will need to move the code to the master branch, and add pointers to the Packagist website, and then test to make sure it is installable from there.

Items after release:
1. Change the templates to move the generated FormObj code into a QPanel instead. The QPanel can then be embedded in a dialog, another panel, or a form. This would then be called a ViewController object, and be more flexible than the current architecture.
2. Move the old plugins to the new plugin architecture that uses Composer for installation.
3. Advertise

If anyone wants to help, the Examples code bug fixes could be started now. Would love to get a sense of how we can integrate @vaibhav-kaushal 's excellent documentation work. And also, someone could start on the ModelController name change.

Into the future:
PSR-1, PSR-2 and PSR-4 with caveats
Templates that generate Hack code (that would make us pretty unique in the PHP world)

**Comment by spekary** [comment]
And thanks to @olegabr for his code reviews. You are an excellent reviewer and catch a lot of things I miss.

**Comment by vaibhav-kaushal** [comment]
@spekary Seriously, that's a cool list! Thanks for seeing through so many changes at one time! It is incredible.

1. I would agree with number 1 point (did you actually mean the issue number 1?)
2. I am trying to use the persistence but it is really not that great - both by usability and by memory footprint. Your approach is more than welcome on this one. 
3. :+1: 
4. Lovely choice for naming! It would have been a problem (and a big one in many cases) to change the name on user level.

About things to do in Beta:

1. I have stopped my efforts to let all changes settle down. Since most of the work on primary controls (which would not change, not significantly at least) are done, I am waiting for the changes to settle down. 
2. Yes, very much needed. Since I do refer them from time to time, it would really help to get them to work perfectly.
3. Really required. Even using only trac is better than current setup. 
4. I have thoughts about this since long. But I think something which can go with travis would be lot better. PHPUnit is preferred almost everywhere for such requirements. 
5. I am working on them. Waiting for the alpha to freeze before making moves and creating things for 2.2.X would be a time wasting mission since some of the basic things will have to be re-done.
6. What do you mean by 'deployment strategy'? Something for end-developers for deploying a QCubed based project into cloud/server easily?

About Items after release: 

1. :+1: I like the idea.
2. :+1: Quite important since I myself am using a bunch of them on my live project(s).
3. Good quality stuff would advertise itself. Of course efforts are needed. 

About my documentation part: api.qcu.be should be updated. PhpDocumenter is by far the best tool for making documentations. We should continue that. I am planning to create some use-case examples but they won't go live anytime soon. About the content on my qcubed.in wiki, I have already said that it is free to be copied on the GitHub wiki.

And yes, @olegabr is a great reviewer for sure!

**Comment by spekary** [comment]
What do you mean by 'deployment strategy'?

Yes, the strategy for deploying from a development environment to a server. Like moving as much as possible out of docroot on the server.

**Comment by olegabr** [comment]
@spekary, I've created separate issues for each your alpha-point: #602, #603, #604, #605 
The 3.0 milestone now have 11 open issues: https://github.com/qcubed/framework/milestones/3.0%20Release

**Comment by vaibhav-kaushal** [comment]
Lovely! That should keep us focussed.

**Comment by vmakre** [comment]
I think we need to put all efforts to finish 3.0 version  . This means someone need to make documentation , rework examples and have good sample startup project to start building  a new application. I want to give my contribution in thinking about easy development and fast application building. 
1. We need to move some part of qcubed library code to plugin part , as sql query watcher (profiler ) why we don't have debugger or sql profiler as plugins only for development  (look at phpdebugbar). There is code generator too we can for example put it in plugin with sql profiler. I am thinking about one good development plugin apart of qcubed like htmlpurifier library, with possibility to manage configuration parameters, generate models and debug sql .So deployment to live server will be easy and secure without this plugin .
2. Write  better documentation about making plugins easy (not science fiction ) but a documentation is a must , plugins are very important , we need html editors, smtp mailer  and a tons of free libraries we can install in one line of command with composer.  There are a lot of questions about static library assets as javascripts and css files in plugins and deployment to live server too .
3.  Rewrite engine and authentication , we need to be clear here plugins or not. How i can make for example authentication plugin without modification of project core files??? do you have some ideas?

**Comment by kmeirlaen** [comment]
As an early qcodo/qcubed contributor (before the github move!), I 'm really excited to see these initiatives. As I want to pick up coding again, I hope to be able to contribute again too! QCubed is still a great framework.
Only issue is :  i need a crash course in github! (svn/csv has been my friend during my coding days).
Anyone up to some resources or a quick personal training?

Also, which branch should I work on? Should I fork the alpha and request merges with master? Sorry for those beginner questions :)

cheers,
Kristof

**Comment by spekary** [comment]
@vmakre and @kmeirlaen, welcome back! Great to see some old contributors.

re: plugins, see the repositories that are labeled plugins_* for current examples. There are 4 or 5 now I think. These are of course just the beginnings of a plugin architecture. Plugins are going to be different than in the past. Some things are better, some not so much, but all-in-all, I think better. 

Plugins install into the vendor/qcubed/plugin directory. At first, this seems awkward, but I think it is best. The old architecture used diffs to actually modify core code, and that just was a bad idea, since if the core changed, the diff would not work. It just was not maintainable. 

For QControl type plugins, a simple composer install is possible. Note that the current plugin controls will be changing a bit in how they implement javascript. I am refactoring the qcbued.js response, and the end result is going to be more compact and easier to implement. But not compatible with the current alpha3 plugins. I will be updating them soon though.

For more complicated plugins, like plugin_bootstrap, there are some post-install things to do. I don't think those things can be done without manually making changes to project/* files, but I think that is OK. There just is no way to predict how different plugins will behave and where they hook in, so for now, having some freedom is good. But, that said, it is possible to create some automation too.

Writing plugins is much easier now than the old process. There is no complicated install config file needed and pre-processor. The only thing that is needed to create is the composer.json file, and there are good examples up now, and its well documented on the composer site. 

One way of creating a kind of automated configuration of plugins is to put some kind of file or extra information in the plugin's folder, and then have qcubed look through those folders. The composer.json file itself can have an "extra" entry that you can add whatever you want (currently used to specify examples for the plugin). Or you can just use a file, like the control_registry.inc.php. See the plugin_phonetextbox repo for examples of both. These are just some things I put together to kind of try out some strategies for automated plugin configuration. Please share your thoughts.

@vmakre's ideas for plugins are great, and all possible. You can inject code into the code generation process either through a plugin template, or creating a class that is a subclass of QCodeGenBase. Tell the developer to change QCodeGen in the project folder to extend from your class, and so your class will be sandwiched between the two classes and add its functionality. That is how the plugin_bootstrap code works. It has its own Control class that sits between QControl and QControlBase. 

Authentication plugin should be pretty straight forward, but if you end up needing changes to the core files, that is OK too. I believe PHP5.5 or 5.6 has some new ways of creating passwords that could be incorporated. Essentially you are creating a login screen, password generation, and then it gets stored in the database. There is nothing wrong with telling people what tables and fields you need in the database for your plugin to work.

@kmeirlaen, yes, git takes some practice. You fork the "qcubed/framework" github repo into your personal github account, and then you will be proposing changes to whatever branch you want, alpha3 being the preferred one. After the fork, you then use git to "clone" your fork to put it on your local computer. You might need to use git to "pull" down the alpha3 branch to your local computer. The next thing to do is to create a new branch based on alpha3, with a name that is related to what you are working on. Once your changes are done, you "commit" them to your local branch on your computer. You then "push" them up to your personal github fork, login to github, and from your local fork, issue a "pull request" on the alpha3 branch. That is when it will get reviewed and merged into the master by one of the core contributors.

**Comment by kmeirlaen** [comment]
Excellent, will give the process a try with some smaller contributions.

**Comment by kmeirlaen** [comment]
I'll need to practice my coding skills - 5 hours later and could not even install QCubed (alpha)
- I don't know Composer
- redirects to incorrect locations (tried /, /install and many more)
- every directory contains a qcubed.inc.php ? 
- dependancy on /vendor directory structure

Tried following installation instructions from install.md but did not become any wizer.

I may be a bit rusty (no coding for over a year now) but if this is the user experience for a first time user, it will turn them down quickly.

For an installer / first page, I would not recommend any redirects, includes, ... the page should simply work and show at least something.

Anyway, i'll try again tomorrow, but installing QCubed 3 really needs some work too.

No criticism, I'm really looking forward to where QCubed is ! Just a bit frustrated that I could not do it today (after a couple of hours trying to get familiar with git and the new QCubed).

**Comment by vmakre** [comment]
@kmeirlaen if you have problem with composer just download qcubed Zipped version . This is good sample project https://github.com/vmakre/qcubed-new-project-sample with bootstrap template and some how to descriptions. There is no database only startup with two pages. When you start coding with qcubed you will never get back. Everything is object oriented and very easy to code and you must have good IDE because of autocomplete i'am  using netbeans with xdebug and xampp.

**Comment by olegabr** [comment]
@kmeirlaen! don't miss the wiki pages: https://github.com/qcubed/framework/wiki
and especially these pages:
  - https://github.com/qcubed/framework/wiki/How-to-git-for-svn-users How to git for svn users
  - https://github.com/qcubed/framework/wiki/Workflow Workflow
  - https://github.com/qcubed/framework/wiki/Roadmap The QCubed Project Roadmap

**Comment by kmeirlaen** [comment]
Thanks all! Still trying. But I think I figured the directory structure out  thanks to this post: http://qcubed.in/qcubed_installation_3_0

**Comment by kmeirlaen** [comment]
Up and Running!

**Comment by spekary** [comment]
Great! We made a decision to move towards a composer install, because it seems to becoming a standard, and is a pretty good dependency manager. Getting Composer installed and running is a bit of a pain, but really important. Once you get that, installing other add-ons and the plugins will be much easier, as all you have to do is add some lines to your composer.json and then run 'composer install'. Its worth the initial pain in the long run.

Once we go to beta, I am thinking about creating some screencasts to walk through the process of installation from start to finish. I have a mac, so if anyone wants to do this on Windows, that would be great.

**Comment by kmeirlaen** [comment]
Thanks. Got composer running and will try to repeat the installation with it later.

**Comment by kmeirlaen** [comment]
Ah, yes, windows here!

**Comment by spekary** [comment]
I am feature complete at this point.

**Comment by vakopian** [comment]
We should have a place to document how to migrate from version 2 to 3. Perhaps a section in README.md? @spekary, if you start an initial list, we can gradually complete it (with pull request to the document), while we port our own applications to version 3.

**Comment by vaibhav-kaushal** [comment]
@spekary There are few more issues remaining ( https://github.com/qcubed/framework/milestones/3.0%20Release ). I think they will not cause any API changes not. I will resume my documentation efforts whenever I get time next. 

@vakopian If you have any suggestions about documentation, I would love to hear it.

**Comment by vakopian** [comment]
@vaibhav-kaushal, I wasn't really talking about full documentation - just about a minimum list of steps to upgrade from v2 to v3, for people who already know QCubed v2.

**Comment by vaibhav-kaushal** [comment]
Since @spekary is the person who has made most of the changes in this release, I think he can lay down the steps and I will try to document it as far as possible. I am waiting for the beta branch to be live in the main repo (which I will wait @spekary to do). 

Speaking of documentation, I myself am aiming for full documentation of the project. I tried doing it in Wiki style and write about features and use-cases etc. but I found that the API documentation needs to be in place first for me to write those. So I am targeting that first. _ And that is what I was asking from you - if you have gone through some of the documentation which has been already written, I’d love any suggestions that you might have._ @olegabr and @spekary here think it is good. Since you are a senior here, your views (and more importantly, any flaws that you detect) can help the project.

**Comment by vakopian** [comment]
I'm just now starting to look at alpha-3 for the first time. Is there a description anywhere of all the new features and large changes that have been made in alpha-3. I searched in github issues and elsewhere, but only found little scattered information. So far I can see "composer" (together with directory structure rework) and the "ModelConnector" renames as large visible changes.
Can anyone provide a list of other significant changes with short descriptions? That would be really helpful and maybe provide the basis for creating the "upgrade instructions" document I mentioned before.

**Comment by vakopian** [comment]
@spekary ,
According to PhpStorm, there are several commits in master that are not in alpha-3. Here is the list with their comments (copied from PhpStorm):
```
98e59278	fix bug introduced in b26cc507daac27597b94503da2eaec99162e694b where jsvascript files added in GetEndScript were not processed in RenderEnd()	Vardan Akopian	2/22/15 1:37 PM
e710e788	Merge pull request #616 from vakopian/preload-js-files	Shannon	2/21/15 6:48 AM
b26cc507	add a configuration option to load the javascript files with <script> tags rather than with qc.loadJavaScriptFile	Vardan Akopian	2/20/15 7:46 PM
98b32f13	Merge pull request #615 from vakopian/validation-reset	Shannon	2/20/15 4:00 PM
746422fb	call the ValidationReset method (which marks controls as Modified) only if the action causes validation. Fixes #581	Vardan Akopian	2/20/15 11:57 AM
c28875b4	Merge pull request #564 from matthiaz/patch-9	Vaibhav Kaushal	12/18/14 1:49 AM
28fe6570	Typo in documentation of generated files	matthiaz	12/18/14 1:39 AM
f7e736b0	fix QAutocomplete options handling	Vardan Akopian	12/12/14 9:28 PM
d1f12be2	fix the old style template generation	Vardan Akopian	12/11/14 2:46 PM
583c92a7	Merge pull request #515 from matthiaz/patch-7	Shannon	12/5/14 2:51 PM
5aed0b15	Merge pull request #534 from matthiaz/patch-8	Oleg Abrosimov	11/21/14 2:20 PM
8d968a37	nl.po Dutch translation of qcubed layout	matthiaz	11/21/14 12:40 AM
085870dc	Strange use of class names in renderwitherror	matthiaz	10/22/14 12:28 AM
```
The top few are probably handled differently in alpha-3, but the bottom 8 might need to be merged.

**Comment by vakopian** [comment]
I guess replying to myself - I went through all the commits in alpha-3 that are not in master, and created a list of "significant" changes in alpha-3. There are tons of fixes and improvements in alpha-3, but I tried to keep the list short and target the parts visible to end-users (developers using QCubed):

* QCubed Installation with composer
* Many-to-many type tables
* QEmailTextBox
* QTabKeyEvent, QBackspaceKeyEvent
* Selector for events
* QWatcher
* Plugins through composer
* Namespaces for plugins
* Use jQuery UI 1.11
* QUrlTextBox
* QFieldset
* Tab names from child controls in QTabs
* QSimpleTable improvements
* QHtmlList
* QCacheProviderAPC
* QDateTime fixes
* Optimistic locking
* QModelTrait
* QMetaParams
* Control Registry
* QCsvTextBox
* QPaginator fixes
* ExecuteJsFunction(), ExecuteSelectorFunction()
* Rename MetaControls to ModelConnectors

It would be great if @spekary and @olegabr who have first hand experiance with alpha-3 could complement/fix/comment on this list

**Comment by spekary** [comment]
Great list, here are my edits based on your list:

PHP 5.4 minimum requirement
All templates are now completely done in PHP, and they all use the <?= tag for better understanding

QCubed Installation with composer, requiring a refactoring of the directory structure. Core code is now in a completely different directory from generated and user code, so its easier to update and maintain, and easier to put in source control.

Many-to-many type tables, giving the same functionality as MySQL's 'SET' type.
QTextBox and derivatives use new PHP validation and filtering commands:
- QEmailTextBox, QUrlTextBox, QCsvTextBox are new derivatives that use this.
Targeting HTML5, UTF-8 charset

QTabKeyEvent, QBackspaceKeyEvent
QOnEvent to trigger actions on bubbled events.
QWatcher to automatically update controls when the database changes.

Use jQuery UI 1.11, jQuery 1.10 or above. No need for compatibility shim.

QFieldset
Tab names from child controls in QTabs

QSimpleTable improvements. Can create spanned cells. 
QHtmlList
QCacheProviderAPC
QDateTime fixes
Optimistic locking?? I don't think anything really changed here.

QModelTrait, including fixes for expanded arrays and significant speed improvements.
QPaginator fixes

ExecuteJsFunction(), ExecuteSelectorFunction() and the beginning of the process to remove the eval in qcubed.js.

Optimizations in qcubed.js, QFormBase, QControlBase and the generated jQueryUI controls for faster Ajax communication and quicker UI updates.

Rename MetaControls to ModelConnectors
~~QMetaParams~~ ModelConnector editor to easily change the type of control created and attributes for that control corresponding to any data field.
Controls generate the ModelConnector parts, rather than the templates. This enables plugins to be able to be styled using the ModelConnector editor described above, and also for sub controls to override how they get created and connected to the database.

Plugins through composer
 - the ability to use namespaces for plugins
 - the ability for a plugin control to be connected to a db object through the ModelConnector (Control Registry helps with this).

QHtml, QHtmlAttributeManager, and the refactoring of the drawing engine. HTML and Javascript can now be produced in a developer mode, with comments, nice formatting and easier javascript debugging, and a minimized mode that is more compact and removes the comments and extra white space. This paves the way for Bootstrap and other CSS framework plugins.

QCheckBox and QRadioButton can produce html that is either wrapped with a label, of has the label next to it.

New Bootstrap plugin

Autocomplete2 plugin for better user experience with the jQueryUI autocomplete.

**Comment by spekary** [comment]
We will need to review those older commits and cherry pick them. Feel free to do so yourself and submit pull requests for the ones you think need to be pulled in to alpha3.

**Comment by vakopian** [comment]
Here we go #629

**Comment by kmeirlaen** [comment]
Very nice list! Looking forward to try it out (was on holidays for the past 1,5 weeks).

**Comment by spekary** [comment]
See https://github.com/qcubed/framework/wiki/Upgrade3 for my initial stab at upgrade instructions.

**Comment by vaibhav-kaushal** [comment]
I believe, we should close the issues with 3.0 label on them which no longer cause a trouble and move closer to the release.

**Comment by vaibhav-kaushal** [comment]
I think the release date is out of question now and we are close anyway. Hence, closing the issue.
